### PR TITLE
Feature/lit 3701 js sdk refactor contract sdk ethers client bootstrapping

### PR DESCRIPTION
**Pull request by joshLong145**
# Description

Refactors `LitContracts` to  allow for optional standup of contract clients  based on the configuration of the `ContractContext`

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by Ansonhkg — COMMENTED** [review]
Suggested change in this PR:
https://github.com/LIT-Protocol/js-sdk/pull/649

File: packages/types/src/lib/types.ts [hunk 1]
```diff
@@ -226,6 +226,21 @@ export interface LitContractContext {
...
+  Pick<LitContractContext, 'LITToken'> &
+  Pick<LitContractContext, 'Multisender'> &
+  Pick<LitContractContext, 'PKPHelper'> &
+  Pick<LitContractContext, 'PKPNFT'> &
+  Pick<LitContractContext, 'PKPNFTMetadata'> &
+  Pick<LitContractContext, 'PKPPermissions'> &
+  Pick<LitContractContext, 'PubkeyRouter'> &
+  Pick<LitContractContext, 'RateLimitNFT'> &
+  Pick<LitContractContext, 'Staking'> &
+  Pick<LitContractContext, 'StakingBalances'>;
```
Comment (FedericoAmura) [inline packages/types/src/lib/types.ts:240]: Why not the other way around?
```suggestion
export interface ExclusiveLitContractContext {
  Allowlist: LitContract;
  LITToken: LitContract;
  Multisender: LitContract;
  PKPHelper: LitContract;
  PKPNFT: LitContract;
  PKPNFTMetadata: LitContract;
  PKPPermissions: LitContract;
  PubkeyRouter: LitContract;
  RateLimitNFT: LitContract;
  Staking: LitContract;
  StakingBalances: LitContract;
}

export interface LitContractContext extends ExclusiveLitContractContext {
  [index: string]: string | any;
}
```

File: packages/contracts-sdk/src/lib/contracts-sdk.ts [hunk 2]
```diff
@@ -422,164 +436,185 @@ export class LitContracts {
...
-        multisenderContract.Multisender,
-    };
+    if (addresses.Allowlist.abi) {
+      this.allowlistContract = {
+        read: new ethers.Contract(
+          addresses.Allowlist.address,
+          addresses.Allowlist.abi as any,
+          this.provider
+        ) as unknown as allowlistContract.Allowlist &
+          allowlistContract.Allowlist,
```
Comment (FedericoAmura) [inline packages/contracts-sdk/src/lib/contracts-sdk.ts:447]:
```suggestion
        ) as allowlistContract.Allowlist,
```

File: packages/contracts-sdk/src/lib/contracts-sdk.ts [hunk 3]
```diff
@@ -422,164 +436,185 @@ export class LitContracts {
...
+        write: new ethers.Contract(
+          addresses.Allowlist.address,
+          addresses.Allowlist.abi as any,
+          this.signer
+        ) as unknown as allowlistContract.Allowlist &
+          allowlistContract.Allowlist,
```
Comment (FedericoAmura) [inline packages/contracts-sdk/src/lib/contracts-sdk.ts:453]:
```suggestion
        ) as allowlistContract.Allowlist,
```

File: packages/contracts-sdk/src/lib/contracts-sdk.ts [hunk 4]
```diff
@@ -422,164 +436,185 @@ export class LitContracts {
...
-        this.signer
-      ) as unknown as pkpHelperContract.PKPHelper & pkpHelperContract.PKPHelper,
-    };
+    if (addresses.LITToken.abi) {
+      this.litTokenContract = {
+        read: new ethers.Contract(
+          addresses.LITToken.address,
+          addresses.LITToken.abi as ethers.ContractInterface,
+          this.provider
+        ) as unknown as litTokenContract.LITToken & litTokenContract.LITToken,
```
Comment (FedericoAmura) [inline packages/contracts-sdk/src/lib/contracts-sdk.ts:463]:
```suggestion
        ) as litTokenContract.LITToken,
```

File: packages/contracts-sdk/src/lib/contracts-sdk.ts [hunk 5]
```diff
@@ -422,164 +436,185 @@ export class LitContracts {
...
+        write: new ethers.Contract(
+          addresses.LITToken.address,
+          addresses.LITToken.abi as ethers.ContractInterface,
+          this.signer
+        ) as unknown as litTokenContract.LITToken & litTokenContract.LITToken,
```
Comment (FedericoAmura) [inline packages/contracts-sdk/src/lib/contracts-sdk.ts:468]:
```suggestion
        ) as litTokenContract.LITToken,
```

File: packages/contracts-sdk/src/lib/contracts-sdk.ts [hunk 6]
```diff
@@ -422,164 +436,185 @@ export class LitContracts {
...
-      ) as unknown as pkpNftContract.PKPNFT & pkpNftContract.PKPNFT,
-    };
+    if (addresses.Multisender.abi) {
+      this.multisenderContract = {
+        read: new ethers.Contract(
+          addresses.Multisender.address,
+          addresses.Multisender.abi as ethers.ContractInterface,
+          this.provider
+        ) as unknown as multisenderContract.Multisender &
+          multisenderContract.Multisender,
```
Comment (FedericoAmura) [inline packages/contracts-sdk/src/lib/contracts-sdk.ts:479]: Same as above

File: packages/contracts-sdk/src/lib/contracts-sdk.ts [hunk 7]
```diff
@@ -422,164 +436,185 @@ export class LitContracts {
...
+        write: new ethers.Contract(
+          addresses.Multisender.address,
+          addresses.Multisender.abi as ethers.ContractInterface,
+          this.signer
+        ) as unknown as multisenderContract.Multisender &
+          multisenderContract.Multisender,
```
Comment (FedericoAmura) [inline packages/contracts-sdk/src/lib/contracts-sdk.ts:485]: Same as above and with the rest of definitions below. I guess this was something required some time ago? Doesn't raise errors on my IDE

File: packages/contracts-sdk/src/lib/contracts-sdk.ts [hunk 8]
```diff
@@ -422,164 +436,185 @@ export class LitContracts {
...
+      };
+    }
+    if (addresses.PKPHelper.abi) {
+      this.pkpHelperContract = {
+        read: new ethers.Contract(
+          addresses.PKPHelper.address,
+          addresses.PKPHelper.abi as any,
```
Comment (FedericoAmura) [inline packages/contracts-sdk/src/lib/contracts-sdk.ts:492]:
```suggestion
          addresses.PKPHelper.abi as ethers.ContractInterface,
```
Same with the other `any`s in `Contract` constructor

File: packages/types/src/lib/types.ts (hunk 1 above)
Comment (Ansonhkg) [reply packages/types/src/lib/types.ts:240]: Yea, I tried that, but the IntelliSense doesn't pick it up. I also tried to omit string from the interface but will give me the following

image

**Review by Ansonhkg — APPROVED** [review]
lgtm now

**Review by FedericoAmura — APPROVED** [review]
Thanks for cleaning those types!

**Pull request merged**